include/bout/mesh.hxx

```diff
@@ -434,17 +435,35 @@ class Mesh {
     ASSERT1(location != CELL_DEFAULT);
     ASSERT1(location != CELL_VSHIFT);
 
-    if (coords_map.count(location)) { // True branch most common, returns immediately
-      return coords_map[location].get();
-    } else {
-      // No coordinate system set. Create default
-      // Note that this can't be allocated here due to incomplete type
-      // (circular dependency between Mesh and Coordinates)
-      coords_map.emplace(location, createDefaultCoordinates(location));
-      return coords_map[location].get();
+#if CHECK > 0
+    if (!coords_map.count(location)) {
+      throw BoutException("Error: Coordinates for %s have not been added to "
+          "this Mesh. You should use REQUEST_STAGGER(location) before "
+          "initializing fields staggered to 'CELL_LOC location'.",
+          CELL_LOC_STRING(location).c_str());
     }
+#endif
+    return coords_map.at(location).get();
   }
 
+  /// Add Coordinates object at a certain location.
+  /// If replace_coords is set to true, reset the object in coords_map if it
+  /// already exists, otherwise add a new one
+  void addCoordinates(const CELL_LOC location, bool replace_coords = false);
+
+  /// Check if Coordinates object at location has been added
+  bool hasCoordinates(const CELL_LOC location) {
+    return coords_map.count(location);
+  }
+
+  /// Count how many Coordinates objects have been added
+  int countCoordinates() {
+    return coords_map.size();
+  }
+
+  /// switch to pass to Coordinates objects
+  bool allow_geometry_without_recalculate_staggered;
```

Comment:
Please could the doxygen comment be a bit more descriptive?

Also should this be a member of `mesh`/does it need to be stored at all? It looks like it should maybe just be an argument to `Coordinates::geometry`?